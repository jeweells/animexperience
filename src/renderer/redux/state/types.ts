import { FStatus, Optional, TopView } from '@shared/types'
import {
  AnimeInfo,
  DeepAnimeIdSearchResult,
  EpisodeInfo,
  FollowedAnime,
  RecentAnimeInfo,
  RecommendationInfo,
  WatchHistoryItem
} from '@shared/types'
import { FollowedAnimeWStatus } from '../reducers/followedAnimes'
import { PreAllowWatch } from '../reducers/invokedLink'
import { RecentAnimeData } from '../../src/hooks/useRecentAnimes'
import { VideoOption } from '../../src/components/VideoPlayer'

export type AnimeSearchState = {
  // Boolean id
  searching?: string
  result?: Optional<DeepAnimeIdSearchResult>
  status: Partial<{
    result: FStatus
    moreResults: FStatus
  }>
}

export type FollowedAnimesState = {
  followedDict: Record<string, FollowedAnime>
  followed: FollowedAnimeWStatus[]
  status: Partial<{
    followed: FStatus
  }>
}

export type InvokedLinkState = {
  open: {
    watch?: string
  }
  preAllow: Partial<{
    watch: PreAllowWatch
  }>
  status: {
    watch?: FStatus
  }
}

export type NotificationsState = {
  message: string | null
  key: string
}

export type PeekState = {
  name?: string
  status: Partial<{
    info: FStatus
  }>
  info?: Optional<AnimeInfo>
  peeking?: string
}

export type PlayerState = {
  open?: string
  freezed?: boolean
}

export type OptionInfo = {
  name: string
  prefer?: boolean
  score: number
}
// Define a type for the slice state
export type PlayerOptionsState = {
  options?: OptionInfo[]
  history?: string[]
  preferred?: string[]
  status: Partial<{
    options: FStatus
  }>
}

export type RecommendationsState = {
  [name: string]: Partial<{
    status: FStatus
    recommendations: RecommendationInfo[]
  }>
}

export type TopviewState = {
  views: TopView[]
}

export type WatchState = {
  watching?: Optional<RecentAnimeData>
  availableVideos?: Optional<VideoOption[]>
  status: {
    availableVideos?: FStatus
    info?: FStatus
    watchEpisode?: FStatus
  }
  info?: Optional<AnimeInfo>
  showNextEpisodeButton?: boolean
  nextEpisodeTimeout: number
  autoFullScreen?: boolean
}

export type WatchedState = {
  episodes: {
    [name: string]: EpisodeInfo
  }
  recently?: RecentAnimeInfo[]
  status: {
    episodes?: FStatus
    recently?: FStatus
  }
}

export type WatchHistoryState = {
  sorted?: WatchHistoryItem[]
  status: {
    sorted?: FStatus
  }
}
