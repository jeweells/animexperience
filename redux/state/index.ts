import {
    AnimeSearchState,
    FollowedAnimesState,
    InvokedLinkState,
    NotificationsState,
    PeekState,
    PlayerOptionsState,
    PlayerState,
    RecommendationsState,
    TopviewState,
    WatchedState,
    WatchHistoryState,
    WatchState,
} from './types'

const animeSearch: AnimeSearchState = {
    status: {},
}

const followedAnimes: FollowedAnimesState = {
    followed: [],
    followedDict: {},
    status: {},
}

const invokedLink: InvokedLinkState = {
    open: {},
    preAllow: {},
    status: {},
}

const notifications: NotificationsState = {
    message: null,
    key: '_',
}

const peek: PeekState = {
    status: {},
}

const player: PlayerState = {}

const playerOptions: PlayerOptionsState = {
    status: {},
}

const recommendations: RecommendationsState = {}

const topView: TopviewState = {
    views: [],
}

const watch: WatchState = {
    status: {},
    nextEpisodeTimeout: -1,
}

const watched: WatchedState = {
    episodes: {},
    status: {},
}

const watchHistory: WatchHistoryState = {
    status: {},
}

export const initialState = {
    animeSearch,
    followedAnimes,
    invokedLink,
    notifications,
    peek,
    player,
    playerOptions,
    recommendations,
    topView,
    watch,
    watched,
    watchHistory,
}

export type RootState = typeof initialState
