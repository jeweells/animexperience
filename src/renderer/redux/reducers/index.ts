export { animeSearch } from './animeSearch'
export { followedAnimes } from './followedAnimes'
export { invokedLink } from './invokedLink'
export { notifications } from './notifications'
export { peek } from './peek'
export { player } from './player'
export { playerOptions } from './playerOptions'
export { recommendations } from './recommendations'
export { topView } from './topView'
export { watch } from './watch'
export { watched } from './watched'
export { watchHistory } from './watchHistory'
