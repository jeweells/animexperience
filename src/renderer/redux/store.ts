import { configureStore } from '@reduxjs/toolkit'
import animeSearch from './reducers/animeSearch'
import followedAnimes from './reducers/followedAnimes'
import peek from './reducers/peek'
import player from './reducers/player'
import playerOptions from './reducers/playerOptions'
import recommendations from './reducers/recommendations'
import topView from './reducers/topView'
import watch from './reducers/watch'
import watched from './reducers/watched'
import watchHistory from './reducers/watchHistory'
import invokedLink from './reducers/invokedLink'
import notifications from './reducers/notifications'

export default configureStore({
  reducer: {
    watch,
    player,
    watched,
    recommendations,
    watchHistory,
    peek,
    topView,
    playerOptions,
    animeSearch,
    followedAnimes,
    invokedLink,
    notifications
  }
})
