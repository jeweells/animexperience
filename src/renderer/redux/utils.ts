import { ThunkDispatch } from 'redux-thunk'
import { RootState } from './state'
import { AnyAction } from '@reduxjs/toolkit'
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux'
import { ForcedAny } from '@shared/types'

export type AppDispatch = ThunkDispatch<RootState, ForcedAny, AnyAction>

export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector

export const useAppDispatch = useDispatch as () => AppDispatch
