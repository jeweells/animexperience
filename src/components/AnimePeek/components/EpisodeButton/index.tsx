import React, { useCallback, useMemo } from 'react'
import styled from 'styled-components'
import { AnimeInfo } from '../../../../../globals/types'
import { watch } from '../../../../../redux/reducers/watch'
import { useAppDispatch } from '../../../../../redux/store'
import { FRowG16 } from '../../../../atoms/Layout'
import { useWatched } from '../../../../hooks/useWatched'
import WatchedRange from '../../../WatchedRange'
import Button from '@mui/material/Button'
import PlayArrowRoundedIcon from '@mui/icons-material/PlayArrowRounded'
export const SEpisodeButton = styled(Button)`
    width: 100%;
    text-align: left;
    justify-content: flex-start;
`

export type EpisodeButtonProps = {
    info: AnimeInfo
    episode: number
}

const WatchedInfo = styled(FRowG16)`
    flex: 1;
    align-items: center;
`

const Content = styled(FRowG16)`
    align-items: center;
    width: 100%;
    justify-content: space-between;
`

export const EpisodeButton: React.FC<EpisodeButtonProps> = React.memo(
    ({ episode, info }) => {
        const epData = useMemo(() => {
            return {
                episode,
                name: info.title,
                link: info.episodeLink.replace(info.episodeReplace, String(episode)),
                img: info.image,
            }
        }, [info, episode])
        const watched = useWatched(epData)
        const dispatch = useAppDispatch()
        return (
            <SEpisodeButton
                onClick={useCallback(() => {
                    dispatch(watch.watchEpisode(epData))
                }, [])}
            >
                <Content>
                    <div>{`Episodio ${episode}`}</div>
                    <WatchedInfo>
                        {watched && (
                            <WatchedRange
                                info={watched}
                                showTime={true}
                                hideBorder={true}
                            />
                        )}
                    </WatchedInfo>
                    <PlayArrowRoundedIcon
                        style={{ fontSize: '1.4rem', overflow: 'hidden' }}
                    />
                </Content>
            </SEpisodeButton>
        )
    },
)

EpisodeButton.displayName = 'EpisodeButton'

export default EpisodeButton
