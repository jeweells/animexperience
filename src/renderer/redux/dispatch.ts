import { AppDispatch } from '~/redux/utils'
import store from '~/redux/store'

export const dispatch: AppDispatch = (action) => store.dispatch(action)
