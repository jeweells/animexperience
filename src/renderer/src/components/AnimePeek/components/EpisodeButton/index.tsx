import { memo, FC, useCallback, useMemo } from 'react'
import { styled } from '@mui/system'
import { AnimeInfo } from '@shared/types'
import { watch } from '@reducers/watch'
import { FRowG16 } from '~/src/atoms/Layout'
import { useWatched } from '~/src/hooks/useWatched'
import WatchedRange from '../../../WatchedRange'
import Button from '@mui/material/Button'
import PlayArrowRoundedIcon from '@mui/icons-material/PlayArrowRounded'
import Tooltip from '@mui/material/Tooltip'
import { EPISODE_BUTTON } from '@selectors'
import { useAppDispatch } from '~/redux/utils'

export const SEpisodeButton = styled(Button)`
  width: 100%;
  text-align: left;
  justify-content: flex-start;
`

export type EpisodeButtonProps = {
  info: AnimeInfo
  episode: number
}

const WatchedInfo = styled(FRowG16)`
  flex: 1;
  align-items: center;
`

const Content = styled(FRowG16)`
  align-items: center;
  width: 100%;
  justify-content: space-between;
`

export const EpisodeButton: FC<EpisodeButtonProps> = memo(({ episode, info }) => {
  const epData = useMemo(() => {
    return {
      episode,
      name: info.title,
      link: info.episodeLink.replace(info.episodeReplace, String(episode)),
      img: info.image
    }
  }, [info, episode])
  const watched = useWatched(epData)
  const dispatch = useAppDispatch()
  return (
    <Tooltip title={'Ver episodio'} arrow placement={'right'}>
      <SEpisodeButton
        data-testid={EPISODE_BUTTON.WATCH_BUTTON}
        onClick={useCallback(() => {
          dispatch(watch.watchEpisode(epData))
        }, [])}
      >
        <Content>
          <div>{`Episodio ${episode}`}</div>
          <WatchedInfo>
            {watched && <WatchedRange info={watched} showTime={true} hideBorder={true} />}
          </WatchedInfo>
          <PlayArrowRoundedIcon style={{ fontSize: '1.4rem', overflow: 'hidden' }} />
        </Content>
      </SEpisodeButton>
    </Tooltip>
  )
})

EpisodeButton.displayName = 'EpisodeButton'

export default EpisodeButton
