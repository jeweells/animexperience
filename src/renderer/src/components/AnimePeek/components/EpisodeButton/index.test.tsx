import { Provider } from 'react-redux'
import EpisodeButton, { EpisodeButtonProps } from './index'
import { RootState } from '~/redux/state'
import { DeepPartial } from 'redux'
import { EPISODE_BUTTON } from '@selectors'
import { MockStoreEnhanced } from 'redux-mock-store'

describe('EpisodeButton', () => {
  let initialState: DeepPartial<RootState>
  let props: EpisodeButtonProps
  let store: MockStoreEnhanced

  const getComponent = (override: Partial<EpisodeButtonProps> = {}) => {
    store = mockStore(initialState)
    return (
      <Provider store={store}>
        <EpisodeButton {...props} {...override} />
      </Provider>
    )
  }

  beforeEach(() => {
    initialState = {
      watched: {
        episodes: {
          69: {
            currentTime: 59,
            at: 101,
            duration: 600
          }
        }
      }
    }

    props = {
      info: {
        description: 'here goes description',
        link: 'here goes link',
        emitted: {
          from: 69,
          to: 129
        },
        related: [
          { name: 'Related name', link: 'related link', type: 'Ova' },
          { name: 'Related name', link: 'related link', type: 'Serie' },
          { name: 'Related name', link: 'related link', type: 'Película' },
          { name: 'Related name', link: 'related link', type: 'Especial' }
        ],
        episodeLink: 'here goes episode link (here goes episode replace)',
        type: 'Serie',
        image: 'here goes image',
        episodeReplace: 'here goes episode replace',
        episodesRange: {
          max: 10,
          min: 1
        },
        otherTitles: ['title1', 'title2'],
        status: 'En emisión',
        tags: ['tag1', 'tag2'],
        thumbnail: 'here goes thumbnail',
        title: 'here goes title'
      },
      episode: 69
    }
  })

  it('renders default', () => {
    const wrapper = render(getComponent())
    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('click to watch should call watchEpisode with expected data', () => {
    const wrapper = render(getComponent())
    fireEvent.click(wrapper.getByTestId(EPISODE_BUTTON.WATCH_BUTTON))
    expect(store.getActions()).toContainEqual(
      expect.objectContaining({
        meta: expect.objectContaining({
          arg: {
            episode: props.episode,
            name: props.info.title,
            link: 'here goes episode link (69)',
            img: props.info.image
          }
        }),
        type: 'watch/watchEpisode/pending'
      })
    )
  })
})
