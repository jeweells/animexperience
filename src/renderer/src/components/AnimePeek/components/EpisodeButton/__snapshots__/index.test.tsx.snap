// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpisodeButton renders default 1`] = `
<DocumentFragment>
  .c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c4 {
  gap: 16px;
}

.c8 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
}

.c0::-moz-focus-inner {
  border-style: none;
}

.c0.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

.c1 {
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.75;
  -webkit-letter-spacing: 0.02857em;
  -moz-letter-spacing: 0.02857em;
  -ms-letter-spacing: 0.02857em;
  letter-spacing: 0.02857em;
  text-transform: uppercase;
  min-width: 64px;
  padding: 6px 8px;
  border-radius: 4px;
  -webkit-transition: background-color 250ms cubic-bezier(0.4,0,0.2,1) 0ms,box-shadow 250ms cubic-bezier(0.4,0,0.2,1) 0ms,border-color 250ms cubic-bezier(0.4,0,0.2,1) 0ms,color 250ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4,0,0.2,1) 0ms,box-shadow 250ms cubic-bezier(0.4,0,0.2,1) 0ms,border-color 250ms cubic-bezier(0.4,0,0.2,1) 0ms,color 250ms cubic-bezier(0.4,0,0.2,1) 0ms;
  color: #1976d2;
}

.c1:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: rgba(25,118,210,0.04);
}

.c1.Mui-disabled {
  color: rgba(0,0,0,0.26);
}

.c7 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: fill 200ms cubic-bezier(0.4,0,0.2,1) 0ms;
  font-size: 1.5rem;
}

.c2 {
  width: 100%;
  text-align: left;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c6 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

@media print {
  .c0 {
    color-adjust: exact;
  }
}

@media (hover:none) {
  .c1:hover {
    background-color: transparent;
  }
}

<button
    aria-label="Ver episodio"
    class="c0 MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root c1 c2"
    data-mui-internal-clone-element="true"
    data-testid="episode-button-watch-button"
    tabindex="0"
    type="button"
  >
    <div
      class="c3 c4 c5"
    >
      <div>
        Episodio 69
      </div>
      <div
        class="c3 c4 c6"
      />
      <svg
        aria-hidden="true"
        class="c7 MuiSvgIcon-root MuiSvgIcon-fontSizeMedium"
        data-testid="PlayArrowRoundedIcon"
        focusable="false"
        style="font-size: 1.4rem; overflow: hidden;"
        viewBox="0 0 24 24"
      >
        <path
          d="M8 6.82v10.36c0 .79.87 1.27 1.54.84l8.14-5.18c.62-.39.62-1.29 0-1.69L9.54 5.98C8.87 5.55 8 6.03 8 6.82z"
        />
      </svg>
    </div>
    <span
      class="c8 MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;
